File: gameVersion2/level_generator.py
"""
level_generator.py
------------------
Procedural level generator.

Every level is a dict:
    {
        "platforms" : [(x, y, w, h), ...],          # canvas-space rects
        "spikes"    : [(x, y, w, h, direction), ...],
        "goal"      : (x, y, w, h),                  # always canvas centre
        "spawn"     : (x, y),                        # always a safe corner
    }

All coordinates are in the 1920×1080 normalised canvas; they are scaled to the
real screen at load time by Game.load_level().
"""

import math
import random

from settings import (
    CANVAS_W, CANVAS_H,
    GOAL_W, GOAL_H, GOAL_CX, GOAL_CY,
    GOAL_SAFE_R, SPAWN_SAFE_R,
    NUM_LEVELS,
)


# ---------------------------------------------------------------------------
# Geometry helpers
# ---------------------------------------------------------------------------

def _overlap(ax, ay, aw, ah, bx, by, bw, bh, margin: int = 20) -> bool:
    """Return True if the two rects (expanded by *margin*) overlap."""
    return (
        ax - margin < bx + bw and ax + aw + margin > bx and
        ay - margin < by + bh and ay + ah + margin > by
    )


def _near_centre(x, y, w, h, safe_r: int = GOAL_SAFE_R) -> bool:
    """Return True if the rect centre is too close to the goal centre."""
    cx = x + w / 2
    cy = y + h / 2
    return math.hypot(cx - GOAL_CX, cy - GOAL_CY) < safe_r


def _near_spawn(x, y, w, h, sx, sy, safe_r: int = SPAWN_SAFE_R) -> bool:
    cx = x + w / 2
    cy = y + h / 2
    return math.hypot(cx - sx, cy - sy) < safe_r


# ---------------------------------------------------------------------------
# Platform placement
# ---------------------------------------------------------------------------

def _place_platform(existing: list, x, y, w, h, sx, sy) -> bool:
    """
    Try to add (x,y,w,h) to *existing*.
    Rejects if it overlaps another platform, is too close to the goal or
    spawn zones, or extends outside the arena margin.
    """
    if _near_centre(x, y, w, h):
        return False
    if _near_spawn(x, y, w, h, sx, sy):
        return False
    margin = 40
    if x < margin or y < margin or x + w > CANVAS_W - margin or y + h > CANVAS_H - margin:
        return False
    for (ex, ey, ew, eh) in existing:
        if _overlap(x, y, w, h, ex, ey, ew, eh):
            return False
    existing.append((x, y, w, h))
    return True


def _add_shape(platforms: list, shape: list, sx, sy) -> bool:
    """
    Attempt to add an entire multi-rect shape atomically.
    If any piece fails placement the whole shape is rejected.
    """
    tmp = list(platforms)
    for piece in shape:
        if not _place_platform(tmp, *piece, sx, sy):
            return False
    platforms.clear()
    platforms.extend(tmp)
    return True


# ---------------------------------------------------------------------------
# Shape generators  (all return a list of (x,y,w,h) tuples)
# ---------------------------------------------------------------------------

def _rand_slab(r: random.Random) -> list:
    """Horizontal slab – the most common building block."""
    w = r.randint(120, 340)
    h = r.randint(18, 28)
    x = r.randint(60, CANVAS_W - 60 - w)
    y = r.randint(80, CANVAS_H - 80 - h)
    return [(x, y, w, h)]


def _rand_pillar(r: random.Random) -> list:
    """Tall vertical pillar."""
    w = r.randint(18, 30)
    h = r.randint(100, 260)
    x = r.randint(60, CANVAS_W - 60 - w)
    y = r.randint(80, CANVAS_H - 80 - h)
    return [(x, y, w, h)]


def _rand_L(r: random.Random) -> list:
    """L-shape: horizontal arm + a vertical leg at one end."""
    w1 = r.randint(160, 280)
    h1 = 22
    w2 = 22
    h2 = r.randint(80, 180)
    x  = r.randint(60, CANVAS_W - 60 - w1)
    y  = r.randint(80, CANVAS_H - 80 - h1 - h2)
    if r.choice([True, False]):
        return [(x, y, w1, h1), (x + w1 - w2, y + h1, w2, h2)]
    return [(x, y, w1, h1), (x, y + h1, w2, h2)]


def _rand_T(r: random.Random) -> list:
    """T-shape: wide horizontal bar + thin vertical stem from the centre."""
    w1 = r.randint(180, 300)
    h1 = 22
    w2 = 22
    h2 = r.randint(60, 140)
    x  = r.randint(60, CANVAS_W - 60 - w1)
    y  = r.randint(80, CANVAS_H - 80 - h1 - h2)
    return [(x, y, w1, h1), (x + w1 // 2 - w2 // 2, y + h1, w2, h2)]


def _rand_staircase(r: random.Random) -> list:
    """3-step diagonal staircase (ascending left-to-right or right-to-left)."""
    step_w = r.randint(90, 150)
    step_h = 20
    dx     = r.choice([-1, 1]) * r.randint(60, 100)
    dy     = r.randint(60, 120)
    x0     = r.randint(200, CANVAS_W - 200 - step_w)
    y0     = r.randint(200, CANVAS_H - 200 - step_h)
    return [(x0 + dx * i, y0 + dy * i, step_w, step_h) for i in range(3)]


def _rand_U(r: random.Random) -> list:
    """U-channel: bottom slab + two vertical side walls."""
    bw     = r.randint(160, 260)
    bh     = 20
    wall_h = r.randint(80, 160)
    wall_w = 20
    x      = r.randint(80, CANVAS_W - 80 - bw)
    y      = r.randint(80, CANVAS_H - 80 - bh - wall_h)
    return [
        (x,              y + wall_h, bw,     bh),      # bottom
        (x,              y,          wall_w, wall_h),   # left wall
        (x + bw - wall_w, y,         wall_w, wall_h),   # right wall
    ]


def _rand_ring(r: random.Random) -> list:
    """Square frame / floating ring (4 thin slabs forming a hollow square)."""
    size = r.randint(160, 260)
    t    = 18
    x    = r.randint(80, CANVAS_W - 80 - size)
    y    = r.randint(80, CANVAS_H - 80 - size)
    return [
        (x,            y,            size,        t),             # top
        (x,            y + size - t, size,        t),             # bottom
        (x,            y + t,        t,           size - 2 * t),  # left
        (x + size - t, y + t,        t,           size - 2 * t),  # right
    ]


def _rand_cross(r: random.Random) -> list:
    """Plus / cross shape: two overlapping bars."""
    arm = r.randint(100, 180)
    t   = 22
    x   = r.randint(100, CANVAS_W - 100 - arm)
    y   = r.randint(100, CANVAS_H - 100 - arm)
    cx2 = x + arm // 2 - t // 2
    cy2 = y + arm // 2 - t // 2
    return [
        (x,   cy2, arm, t),    # horizontal arm
        (cx2, y,   t,   arm),  # vertical arm
    ]


# Shape function registry (slabs are weighted 3× because they are most useful)
_SHAPE_FUNCS = [
    _rand_slab, _rand_slab, _rand_slab,
    _rand_pillar,
    _rand_L,
    _rand_T,
    _rand_staircase,
    _rand_U,
    _rand_ring,
    _rand_cross,
]


# ---------------------------------------------------------------------------
# Spike generator
# ---------------------------------------------------------------------------

def _spikes_for_platform(
    px, py, pw, ph,
    difficulty: int,
    r: random.Random,
) -> list:
    """
    Return 0–3 spike tuples for one platform rect.
    Each tuple: (x, y, w, h, direction)
    *difficulty* (0–4) controls face count and placement probability.
    """
    spike_t = 24   # spike "thickness" (height for up/down, width for left/right)

    # All four possible spike faces
    faces = [
        ("up",    px,            py - spike_t,  pw,      spike_t),
        ("down",  px,            py + ph,       pw,      spike_t),
        ("right", px + pw,       py,            spike_t, ph),
        ("left",  px - spike_t,  py,            spike_t, ph),
    ]
    r.shuffle(faces)

    max_faces = 1 + difficulty // 2
    prob      = 0.35 + difficulty * 0.10

    spikes = []
    for direction, sx, sy, sw, sh in faces[:max_faces]:
        if r.random() > prob:
            continue
        # Clamp inside arena
        sx = max(10, min(CANVAS_W - 10 - sw, sx))
        sy = max(10, min(CANVAS_H - 10 - sh, sy))
        spikes.append((sx, sy, sw, sh, direction))

    return spikes


# ---------------------------------------------------------------------------
# Top-level generator
# ---------------------------------------------------------------------------

def generate_level(level_idx: int, seed: int = None) -> dict:
    """
    Build and return a complete level dict.

    The goal is always at the exact canvas centre.
    The spawn is always in one of four safe corner regions.
    Difficulty scales 0 → 4 over the first five levels and stays at 4 beyond.
    """
    r          = random.Random(seed if seed is not None else level_idx * 7919 + 42)
    difficulty = min(level_idx, 4)

    # --- Goal (always centre) ---
    goal = (GOAL_CX - GOAL_W // 2, GOAL_CY - GOAL_H // 2, GOAL_W, GOAL_H)

    # --- Spawn (random safe corner) ---
    corners = [(80, 80), (1840, 80), (80, 1000), (1840, 1000)]
    spawn   = r.choice(corners)

    # --- Platforms ---
    platforms: list = []

    # One guaranteed border edge so there is always a landing surface
    # regardless of gravity direction.
    border = r.choice(["floor", "ceiling", "left", "right"])
    border_rects = {
        "floor":   (0,    1050, CANVAS_W,      30),
        "ceiling": (0,    0,    CANVAS_W,      30),
        "left":    (0,    0,    30,        CANVAS_H),
        "right":   (1890, 0,    30,        CANVAS_H),
    }
    platforms.append(border_rects[border])

    # Guaranteed safe landing pad right at spawn
    pad_w, pad_h = 200, 22
    pad_x = max(10, min(CANVAS_W - 10 - pad_w, spawn[0] - pad_w // 2))
    pad_y = max(10, min(CANVAS_H - 10 - pad_h, spawn[1] + 30))
    platforms.append((pad_x, pad_y, pad_w, pad_h))

    # Fill in random shapes
    target = 10 + difficulty * 2
    attempts = 0
    while len(platforms) < target + 2 and attempts < 400:
        attempts += 1
        shape = r.choice(_SHAPE_FUNCS)(r)
        _add_shape(platforms, shape, spawn[0], spawn[1])

    # --- Spikes ---
    blacklist = [
        (GOAL_CX  - GOAL_SAFE_R,  GOAL_CY  - GOAL_SAFE_R,
         GOAL_SAFE_R  * 2,        GOAL_SAFE_R  * 2),
        (spawn[0] - SPAWN_SAFE_R, spawn[1] - SPAWN_SAFE_R,
         SPAWN_SAFE_R * 2,        SPAWN_SAFE_R * 2),
    ]

    spikes: list = []
    for px, py, pw, ph in platforms[2:]:   # skip border + spawn pad
        for sx, sy, sw, sh, direction in _spikes_for_platform(px, py, pw, ph, difficulty, r):
            blocked = any(
                _overlap(sx, sy, sw, sh, bx, by, bw, bh, margin=0)
                for bx, by, bw, bh in blacklist
            )
            if not blocked:
                spikes.append((sx, sy, sw, sh, direction))

    return {
        "platforms": platforms,
        "spikes":    spikes,
        "goal":      goal,
        "spawn":     spawn,
    }


# Pre-generate all levels at import time so Game can index straight into this list.
LEVELS_DATA: list = [generate_level(i) for i in range(NUM_LEVELS)]

File: gameVersion7Android/src/main/kotlin/com/chaosball/game/LevelGenerator.kt
package com.chaosball.game

import kotlin.math.*
import kotlin.random.Random

/** Level Generator - Handles procedural level creation */
object LevelGenerator {

    private fun overlap(ax: Float, ay: Float, aw: Float, ah: Float,
                        bx: Float, by: Float, bw: Float, bh: Float, margin: Float = 20f) =
        ax - margin < bx + bw && ax + aw + margin > bx &&
        ay - margin < by + bh && ay + ah + margin > by

    private fun nearCentre(x: Float, y: Float, w: Float, h: Float, safeR: Float = GOAL_SAFE_R): Boolean {
        val dx = x + w / 2f - GOAL_CX; val dy = y + h / 2f - GOAL_CY
        return sqrt(dx * dx + dy * dy) < safeR
    }

    private fun nearSpawn(x: Float, y: Float, w: Float, h: Float, sx: Float, sy: Float): Boolean {
        val dx = x + w / 2f - sx; val dy = y + h / 2f - sy
        return sqrt(dx * dx + dy * dy) < SPAWN_SAFE_R
    }

    private fun placePlatform(existing: MutableList<Rect>, x: Float, y: Float, w: Float, h: Float, sx: Float, sy: Float): Boolean {
        if (nearCentre(x, y, w, h)) return false
        if (nearSpawn(x, y, w, h, sx, sy)) return false
        val mg = 40f
        if (x < mg || y < mg || x + w > CANVAS_W - mg || y + h > CANVAS_H - mg) return false
        for (e in existing) if (overlap(x, y, w, h, e.x, e.y, e.w, e.h)) return false
        existing.add(Rect(x, y, w, h)); return true
    }

    private fun addShape(platforms: MutableList<Rect>, shape: List<Rect>, sx: Float, sy: Float): Boolean {
        val tmp = platforms.toMutableList()
        for (p in shape) if (!placePlatform(tmp, p.x, p.y, p.w, p.h, sx, sy)) return false
        platforms.clear(); platforms.addAll(tmp); return true
    }

    // ── Shape generators ──────────────────────────────────────────────────

    private fun randSlab(r: Random): List<Rect> {
        val w = r.nextInt(120, 341).toFloat(); val h = r.nextInt(18, 29).toFloat()
        return listOf(Rect(r.nextInt(60, (CANVAS_W - 60 - w).toInt()).toFloat(),
                           r.nextInt(80, (CANVAS_H - 80 - h).toInt()).toFloat(), w, h))
    }

    private fun randPillar(r: Random): List<Rect> {
        val w = r.nextInt(18, 31).toFloat(); val h = r.nextInt(100, 261).toFloat()
        return listOf(Rect(r.nextInt(60, (CANVAS_W - 60 - w).toInt()).toFloat(),
                           r.nextInt(80, (CANVAS_H - 80 - h).toInt()).toFloat(), w, h))
    }

    private fun randL(r: Random): List<Rect> {
        val w1 = r.nextInt(160, 281).toFloat(); val h1 = 22f; val w2 = 22f; val h2 = r.nextInt(80, 181).toFloat()
        val x = r.nextInt(60, (CANVAS_W - 60 - w1).toInt()).toFloat()
        val y = r.nextInt(80, (CANVAS_H - 80 - h1 - h2).toInt()).toFloat()
        return if (r.nextBoolean()) listOf(Rect(x,y,w1,h1), Rect(x+w1-w2, y+h1, w2, h2))
        else listOf(Rect(x,y,w1,h1), Rect(x, y+h1, w2, h2))
    }

    private fun randT(r: Random): List<Rect> {
        val w1 = r.nextInt(180, 301).toFloat(); val h1 = 22f; val w2 = 22f; val h2 = r.nextInt(60, 141).toFloat()
        val x = r.nextInt(60, (CANVAS_W - 60 - w1).toInt()).toFloat()
        val y = r.nextInt(80, (CANVAS_H - 80 - h1 - h2).toInt()).toFloat()
        return listOf(Rect(x,y,w1,h1), Rect(x+w1/2f-w2/2f, y+h1, w2, h2))
    }

    private fun randStaircase(r: Random): List<Rect> {
        val sw = r.nextInt(90, 151).toFloat(); val sh = 20f
        val dx = (if (r.nextBoolean()) 1 else -1) * r.nextInt(60, 101).toFloat()
        val dy = r.nextInt(60, 121).toFloat()
        val x0 = r.nextInt(200, (CANVAS_W - 200 - sw).toInt()).toFloat()
        val y0 = r.nextInt(200, (CANVAS_H - 200 - sh).toInt()).toFloat()
        return (0..2).map { Rect(x0 + dx * it, y0 + dy * it, sw, sh) }
    }

    private fun randU(r: Random): List<Rect> {
        val bw = r.nextInt(160, 261).toFloat(); val bh = 20f
        val wh = r.nextInt(80, 161).toFloat(); val ww = 20f
        val x = r.nextInt(80, (CANVAS_W - 80 - bw).toInt()).toFloat()
        val y = r.nextInt(80, (CANVAS_H - 80 - bh - wh).toInt()).toFloat()
        return listOf(Rect(x, y+wh, bw, bh), Rect(x, y, ww, wh), Rect(x+bw-ww, y, ww, wh))
    }

    private fun randRing(r: Random): List<Rect> {
        val sz = r.nextInt(160, 261).toFloat(); val t = 18f
        val x = r.nextInt(80, (CANVAS_W - 80 - sz).toInt()).toFloat()
        val y = r.nextInt(80, (CANVAS_H - 80 - sz).toInt()).toFloat()
        return listOf(Rect(x,y,sz,t), Rect(x,y+sz-t,sz,t), Rect(x,y+t,t,sz-2*t), Rect(x+sz-t,y+t,t,sz-2*t))
    }

    private fun randCross(r: Random): List<Rect> {
        val arm = r.nextInt(100, 181).toFloat(); val t = 22f
        val x = r.nextInt(100, (CANVAS_W - 100 - arm).toInt()).toFloat()
        val y = r.nextInt(100, (CANVAS_H - 100 - arm).toInt()).toFloat()
        return listOf(Rect(x, y+arm/2f-t/2f, arm, t), Rect(x+arm/2f-t/2f, y, t, arm))
    }

    private fun shapes(r: Random): List<Rect> = when (r.nextInt(10)) {
        0,1,2 -> randSlab(r); 3 -> randPillar(r); 4 -> randL(r)
        5 -> randT(r); 6 -> randStaircase(r); 7 -> randU(r)
        8 -> randRing(r); else -> randCross(r)
    }

    private fun spikesForPlatform(px: Float, py: Float, pw: Float, ph: Float, difficulty: Int, r: Random): List<Spike> {
        val st = 24f
        val faces = listOf(
            Pair(SpikeDir.UP,    Rect(px, py-st, pw, st)),
            Pair(SpikeDir.DOWN,  Rect(px, py+ph, pw, st)),
            Pair(SpikeDir.RIGHT, Rect(px+pw, py, st, ph)),
            Pair(SpikeDir.LEFT,  Rect(px-st, py, st, ph))
        ).shuffled(r)
        val maxF = 1 + difficulty / 2
        val prob = 0.35f + difficulty * 0.10f
        val out = mutableListOf<Spike>()
        for ((dir, rect) in faces.take(maxF)) {
            if (r.nextFloat() > prob) continue
            val sx = rect.x.coerceIn(10f, CANVAS_W - 10f - rect.w)
            val sy = rect.y.coerceIn(10f, CANVAS_H - 10f - rect.h)
            out.add(Spike(Rect(sx, sy, rect.w, rect.h), dir))
        }
        return out
    }

    private fun genRotators(count: Int, spawn: Vec2, r: Random): List<RotatorData> {
        val out = mutableListOf<RotatorData>(); var att = 0
        while (out.size < count && att < 200) {
            att++
            val px = r.nextInt(200, (CANVAS_W - 200).toInt()).toFloat()
            val py = r.nextInt(150, (CANVAS_H - 150).toInt()).toFloat()
            if (hypot((px - GOAL_CX).toDouble(), (py - GOAL_CY).toDouble()) < 200) continue
            if (hypot((px - spawn.x).toDouble(), (py - spawn.y).toDouble()) < 220) continue
            val arm = r.nextInt(90, 161).toFloat()
            val thick = r.nextInt(14, 23).toFloat()
            val speed = (if (r.nextBoolean()) 1 else -1) * r.nextFloat() * 40f + 40f
            out.add(RotatorData(Vec2(px, py), arm, thick, speed))
        }
        return out
    }

    private fun genEnemies(count: Int, spawn: Vec2, r: Random): List<EnemyData> {
        val corners = listOf(Vec2(120f,120f), Vec2(1800f,120f), Vec2(120f,960f), Vec2(1800f,960f))
            .sortedByDescending { it.dist(spawn) }
        return corners.take(minOf(count, corners.size)).map {
            EnemyData(Vec2(it.x + r.nextInt(-60,61).toFloat(), it.y + r.nextInt(-60,61).toFloat()),
                      r.nextFloat() * 45f + 65f)
        }
    }

    private fun genShooters(count: Int, spawn: Vec2, r: Random): List<ShooterData> {
        val out = mutableListOf<ShooterData>(); var att = 0
        while (out.size < count && att < 200) {
            att++
            val px = r.nextInt(200, (CANVAS_W - 200).toInt()).toFloat()
            val py = r.nextInt(150, (CANVAS_H - 150).toInt()).toFloat()
            if (hypot((px - GOAL_CX).toDouble(), (py - GOAL_CY).toDouble()) < 220) continue
            if (hypot((px - spawn.x).toDouble(), (py - spawn.y).toDouble()) < 240) continue
            out.add(ShooterData(Vec2(px, py), r.nextFloat() * 1.7f + 1.8f))
        }
        return out
    }

    private fun genVoidWalls(r: Random) = listOf(
        VoidWallData(VoidOrientation.VERTICAL,   if (r.nextBoolean()) VoidSide.NEAR else VoidSide.FAR),
        VoidWallData(VoidOrientation.HORIZONTAL, if (r.nextBoolean()) VoidSide.NEAR else VoidSide.FAR)
    )

    private fun genCoins(platforms: List<Rect>, spawn: Vec2, difficulty: Int, r: Random): List<Vec2> {
        val coins = mutableListOf<Vec2>()
        val count = 4 + difficulty * 2
        for (p in platforms.drop(2)) {
            if (coins.size >= count) break
            val n = r.nextInt(1, minOf(4, maxOf(2, (p.w / 120f).toInt() + 1)))
            for (i in 0 until n) {
                val cx = p.x + p.w * (i + 1f) / (n + 1f)
                val cy = p.y - 40f
                if (hypot((cx - GOAL_CX).toDouble(), (cy - GOAL_CY).toDouble()) < GOAL_SAFE_R + 30) continue
                if (hypot((cx - spawn.x).toDouble(), (cy - spawn.y).toDouble()) < SPAWN_SAFE_R) continue
                if (cx < 40 || cx > CANVAS_W - 40 || cy < 40 || cy > CANVAS_H - 40) continue
                coins.add(Vec2(cx, cy))
                if (coins.size >= count) break
            }
        }
        return coins
    }

    private fun genBoostPads(platforms: List<Rect>, spawn: Vec2, r: Random): List<Triple<Float,Float,String>> {
        val dirs = listOf("up","left","right")
        val cands = platforms.drop(2).filter { it.w > 100f }.toMutableList().also { it.shuffle(r) }
        val out = mutableListOf<Triple<Float,Float,String>>()
        for (p in cands.take(3)) {
            val cx = p.x + p.w / 2f; val cy = p.y - 1f
            if (hypot((cx-GOAL_CX).toDouble(),(cy-GOAL_CY).toDouble()) < GOAL_SAFE_R+50) continue
            if (hypot((cx-spawn.x).toDouble(),(cy-spawn.y).toDouble()) < SPAWN_SAFE_R) continue
            out.add(Triple(cx, cy, dirs[r.nextInt(dirs.size)]))
        }
        return out
    }

    fun generate(levelIdx: Int): LevelData {
        val seed = levelIdx * 7919L + 42L
        val r = Random(seed); val difficulty = minOf(levelIdx, 4)
        val isSpinner = levelIdx in TIER_BASIC_END until TIER_SPINNER_END
        val isVoid    = levelIdx in TIER_SPINNER_END until TIER_VOID_END
        val isShooter = levelIdx >= TIER_SHOOTER_START

        val goal = Rect(GOAL_CX - GOAL_W/2f, GOAL_CY - GOAL_H/2f, GOAL_W, GOAL_H)
        val spawnChoices = listOf(Vec2(80f,80f), Vec2(1840f,80f), Vec2(80f,1000f), Vec2(1840f,1000f))
        val spawn = spawnChoices[r.nextInt(spawnChoices.size)]

        val platforms = mutableListOf<Rect>()
        val boundaryOptions = listOf(Rect(0f,1050f,CANVAS_W,30f), Rect(0f,0f,CANVAS_W,30f), Rect(0f,0f,30f,CANVAS_H), Rect(1890f,0f,30f,CANVAS_H))
        platforms.add(boundaryOptions[r.nextInt(boundaryOptions.size)])
        val pw2 = 200f; val ph2 = 22f; val px2 = (spawn.x - pw2/2f).coerceIn(10f, CANVAS_W - 10f - pw2); val py2 = (spawn.y + 30f).coerceIn(10f, CANVAS_H - 10f - ph2)
        platforms.add(Rect(px2, py2, pw2, ph2))

        var att = 0; while (platforms.size < 10 + difficulty * 2 + 2 && att < 400) { att++; addShape(platforms, shapes(r), spawn.x, spawn.y) }

        val bl = listOf(Rect(GOAL_CX-GOAL_SAFE_R, GOAL_CY-GOAL_SAFE_R, GOAL_SAFE_R*2f, GOAL_SAFE_R*2f), Rect(spawn.x-SPAWN_SAFE_R, spawn.y-SPAWN_SAFE_R, SPAWN_SAFE_R*2f, SPAWN_SAFE_R*2f))
        val spikes = mutableListOf<Spike>()
        for (p in platforms.drop(2)) {
            for (s in spikesForPlatform(p.x, p.y, p.w, p.h, difficulty, r)) {
                if (bl.none { b -> overlap(s.rect.x,s.rect.y,s.rect.w,s.rect.h, b.x,b.y,b.w,b.h, 0f) }) spikes.add(s)
            }
        }

        val rotators = if (isSpinner) genRotators(listOf(1,2,2,3)[levelIdx - TIER_BASIC_END], spawn, r) else if (isVoid) genRotators(r.nextInt(1,3), spawn, r) else emptyList()
        val enemies = if (isVoid) genEnemies(if (isShooter) r.nextInt(2,4) else r.nextInt(1,3), spawn, r) else emptyList()
        val shooters = if (isShooter) genShooters(listOf(1,2,2,3)[levelIdx - TIER_SHOOTER_START], spawn, r) else emptyList()
        val voidWalls = if (isVoid) genVoidWalls(r) else emptyList()
        val coins = genCoins(platforms, spawn, difficulty, r); val boostPads = if (levelIdx >= 4) genBoostPads(platforms, spawn, r) else emptyList()

        return LevelData(platforms, spikes, goal, spawn, rotators, enemies, shooters, voidWalls, isVoid, coins, boostPads)
    }

    val ALL_LEVELS: List<LevelData> by lazy { (0 until NUM_LEVELS).map { generate(it) } }
}
